### PR TITLE
Unify the metrics implementation between low-level and high-level API.

### PR DESCRIPTION
**Pull request by qingqing01**
### PR types

Function optimization

### PR changes

APIs

### Describe

1. Unify the metrics implementation between low-level and high-level API.
2. Change function name `add_metric_op`  in Metric to `compute`.
3. Return scalar, not  [scalar] in `update()` and `accumulate()` of `Accuracy`.
4.  Refine and add comments for sub-function in `Accuracy`.
    -   Add comments for `compute`, `update`, `reset`..
    -   Add standalone usage and usage with `Model`  API for Metrics
5. Add unit testing in `python/paddle/tests`

### CODE REVIEW

**Review by LielinJiang — DISMISSED** [review]

File: python/paddle/metric/metrics.py [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+    call :code:`update` with states in NumPy format.
+    Metric calculated as follows (operations in Model and Metric are
+    indicated with curly brackets, while data nodes not):
+                 inputs & labels              || ------------------
+                       |                      ||
+                    {model}                   ||
+                       |                      ||
+                outputs & labels              ||
+                       |                      ||    tensor data
+             {Metric.compute}                 ||
```
Comment (heavengate) [inline python/paddle/metric/metrics.py]: 可以调一下缩进对齐一下~

Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

**Review by heavengate — DISMISSED** [review]

File: python/paddle/metric/metrics.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+
+    Refer to:
+    https://en.wikipedia.org/wiki/Precision_and_recall
+
+    Noted that this class manages the recall score only for
+    binary classification task.
+
+    Args:
+        name (str, optional): String name of the metric instance.
+            Default is `precision`.
```
Comment (guoshengCS) [inline python/paddle/metric/metrics.py]: 这里name默认值是否应是`recall`呢

File: python/paddle/incubate/hapi/model.py [hunk 3]
```diff
@@ -1059,15 +1059,15 @@ def forward(self, x):
...
         """
         Configures the model before runing.
 
         Args:
             optimizer (Optimizer|None): Optimizer must be set in training
                 and should be a Optimizer instance. It can be None in eval
                 and test mode.
-            loss_function (Loss|callable function|None): Loss function can
+            loss (Loss|callable function|None): Loss function can
                 be a `fluid.dygraph.Layer` instance or any callable function
```
Comment (saxon-zh) [inline python/paddle/incubate/hapi/model.py]: fluid.dygraph.Layer => paddle.nn.Layer ，下面用到fluid.dygraph.Layer的地方同样都需要修改一下

File: python/paddle/metric/metrics.py [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+    `true_negatives`, `false_positives` and `false_negatives` that are used to
+    compute the AUC. To discretize the AUC curve, a linearly spaced set of
+    thresholds is used to compute pairs of recall and precision values. The area
+    under the ROC-curve is therefore computed using the height of the recall
+    values by the false positive rate, while the area under the PR-curve is the
+    computed using the height of the precision values by the recall.
+
+    Args:
+        name (str, optional): String name of the metric instance. Default
+            is `acc`.
```
Comment (guoshengCS) [inline python/paddle/metric/metrics.py]: 这里name默认值是否应是 `auc` 呢

File: python/paddle/incubate/hapi/model.py [hunk 5]
```diff
@@ -1362,15 +1362,15 @@ def evaluate(
...
             label = hapi.Input('label', [None, 1], 'int64')
             model = hapi.Model(hapi.vision.LeNet(), input, label)
-            model.prepare(metrics=hapi.metrics.Accuracy())
+            model.prepare(metrics=paddle.metric.Accuracy())
 
             result = model.evaluate(val_dataset, batch_size=64)
             print(result)
 
             # imperative mode
             fluid.enable_dygraph()
```
Comment (saxon-zh) [inline python/paddle/incubate/hapi/model.py]: fluid.enable_dygraph => paddle.disable_static

File: python/paddle/metric/metrics.py [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+        curve (str): Specifies the mode of the curve to be computed,
+            'ROC' or 'PR' for the Precision-Recall-curve. Default is 'ROC'.
```
Comment (guoshengCS) [inline python/paddle/metric/metrics.py:585]: Maybe add doc for `num_thresholds` later

**Review by guoshengCS — DISMISSED** [review]

File: python/paddle/metric/metrics.py [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+        shape is [N, 10], :code:`label` shape is [N, 1], N is mini-batch size,
+        and we only need to calculate accurary of top-1 and top-5, we could
+        calculate the correct prediction matrix of the top-5 scores of the
+        prediction of each sample like follows, while the correct prediction
+        matrix shape is [N, 5].
+
+        .. code-block:: python
+            def compute(pred, label):
+                # sort prediction and slice the top-5 scores
+                pred = fluid.layers.argsort(pred, descending=True)[1][:, :5]
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: argsort用paddle.下面的API，去掉fluid

File: python/paddle/metric/metrics.py [hunk 8]
```diff
@@ -0,0 +1,733 @@
...
+                # calculate whether the predictions are correct
+                correct = pred == label
+                return fluid.layers.cast(correct, dtype='float32')
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: 去掉fluid

File: python/paddle/metric/metrics.py [hunk 9]
```diff
@@ -0,0 +1,733 @@
...
+
+        see :code:`Metric.update`
+        """
+        return args
+
+
+class Accuracy(Metric):
+    """
+    Encapsulates accuracy metric logic.
+    def __init__(self, topk=(1, ), name='acc', *args, **kwargs):
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: 没用的信息吧？

File: python/paddle/metric/metrics.py [hunk 10]
```diff
@@ -0,0 +1,733 @@
...
+
+    Example by standalone:
+        
+        .. code-block:: python
+
+        import numpy as np
+        import paddle
+
+        paddle.disable_static()
+        x = paddle.to_variable(np.array([
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: to_variable => to_tensor

File: python/paddle/metric/metrics.py [hunk 11]
```diff
@@ -0,0 +1,733 @@
...
+            [0.1, 0.2, 0.3, 0.4],
+            [0.1, 0.4, 0.3, 0.2],
+            [0.1, 0.2, 0.4, 0.3],
+            [0.1, 0.2, 0.3, 0.4]]))
+        y = paddle.to_variable(np.array([[0], [1], [2], [3]]))
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: 同上

File: python/paddle/metric/metrics.py [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+        res = m.accumulate()
+        print(res) # 0.75
+
+
+    Example with Model API:
+        
+        .. code-block:: python
+
+        import paddle
+        import paddle.fluid as fluid
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: 去掉fluid

File: python/paddle/metric/metrics.py [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+        import paddle.incubate.hapi as hapi
+
+        paddle.disable_static()
+        train_dataset = hapi.datasets.MNIST(mode='train')
+
+        model = hapi.Model(hapi.vision.LeNet(classifier_activation=None))
+        optim = fluid.optimizer.Adam(
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: paddle.optimizer.Adam

File: python/paddle/metric/metrics.py (hunk 4 above)
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: acc => auc

File: python/paddle/metric/metrics.py (hunk 2 above)
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: precision => recall

File: python/paddle/metric/metrics.py [hunk 14]
```diff
@@ -0,0 +1,733 @@
...
+        
+        paddle.disable_static()
+        model = hapi.Model(nn.Sequential(
+            nn.Linear(10, 2, act='softmax'),
+        ))
+        optim = paddle.optimizer.Adam(
+            learning_rate=0.001, parameter_list=model.parameters())
+        
+        def loss(x, y):
+            return fluid.layers.cross_entropy(x, y)
```
Comment (saxon-zh) [inline python/paddle/metric/metrics.py]: 去fluid

File: python/paddle/metric/metrics.py (hunk 8 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done. Thx!

File: python/paddle/metric/metrics.py (hunk 7 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done. Thx!

File: python/paddle/metric/metrics.py (hunk 9 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Removed. Thanks!

File: python/paddle/metric/metrics.py (hunk 10 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

File: python/paddle/metric/metrics.py (hunk 11 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

File: python/paddle/metric/metrics.py (hunk 12 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

File: python/paddle/metric/metrics.py (hunk 13 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

File: python/paddle/metric/metrics.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done.

Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

File: python/paddle/metric/metrics.py (hunk 6 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py:585]: Done.

File: python/paddle/metric/metrics.py (hunk 4 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

Comment (qingqing01) [reply python/paddle/metric/metrics.py]: Done

File: python/paddle/metric/metrics.py (hunk 14 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: done

File: python/paddle/metric/__init__.py [hunk 15]
```diff
@@ -12,17 +12,16 @@
...
-# TODO: define the functions to calculate metric in this directory 
-__all__ = [
-    'Accuracy', 'Auc', 'ChunkEvaluator', 'CompositeMetric', 'DetectionMAP',
-    'EditDistance', 'Precision', 'Recall', 'accuracy', 'auc', 'chunk_eval',
-    'cos_sim', 'mean_iou'
-]
-
-
-
-from ..fluid.metrics import Accuracy, Auc, ChunkEvaluator, CompositeMetric, DetectionMAP, EditDistance, \
```
Comment (jzhang533) [inline python/paddle/metric/__init__.py:24]: paddle/fluid/metrics.py 里的API计划怎么处理？
还有`ChunkEvaluator, CompositeMetric, DetectionMAP, EditDistance`在重构后的paddle.metric下没有对应的API。

File: python/paddle/metric/__init__.py [hunk 16]
```diff
@@ -12,17 +12,16 @@
...
-        Precision, Recall
+from .metrics import *
+from . import metrics
 
 from ..fluid.layers.metric_op import accuracy, auc
 from ..fluid.layers.nn import chunk_eval, cos_sim, mean_iou
```
Comment (jzhang533) [inline python/paddle/metric/__init__.py:19]: paddle.metric下的class形式API，跟小写的调用op的API的关系，是否讨论过啊？

File: python/paddle/metric/metrics.py [hunk 17]
```diff
@@ -0,0 +1,738 @@
...
+        model = hapi.Model(nn.Sequential(
+            nn.Linear(10, 1),
+            nn.Sigmoid()
+        ))
+        optim = paddle.optimizer.Adam(
+            learning_rate=0.001, parameter_list=model.parameters())
+        model.prepare(
+            optim,
+            loss=nn.BCELoss(),
+            metrics=paddle.metric.Recall())
```
Comment (jzhang533) [inline python/paddle/metric/metrics.py]: 这里给一个可以同时计算precison和recall的例子？
通常情况下会同时算这两个指标的。

File: tools/wlist.json [hunk 18]
```diff
@@ -107,12 +107,27 @@
...
+        "Recall.reset",
+        "Recall.update",
+        "Recall.accumulate",
+        "Recall.name",
+        "Recall.compute",
+        "Auc.reset",
+        "Auc.update",
+        "Auc.accumulate",
+        "Auc.name",
+        "Auc.compute",
```
Comment (jzhang533) [inline tools/wlist.json:129]: 同意这次对于白名单的改动。因为在class Auc等的地方已经有了示例代码。

File: python/paddle/metric/__init__.py (hunk 15 above)
Comment (jzhang533) [reply python/paddle/metric/__init__.py:24]: @XiaoguangHu01 晓光麻烦看看__init__.py文件的改动。

Comment (qingqing01) [reply python/paddle/metric/__init__.py:24]: 相关被使用统计在agroup中。 DetectionMAP、EditDistance建议先废弃，相关套件已在模型里重新实现。Precision/Recall当前没有被任何模型使用到，当前fluid下面的实现还太简单。 会上讨论，本次PR先以统一位置为主，后续版本还需要增强相关功能，包括增加ChunkEvaluator,以及改进功能。

File: python/paddle/metric/__init__.py (hunk 16 above)
Comment (qingqing01) [reply python/paddle/metric/__init__.py:19]: class形式:   提供batch累积功能。一般用于高层API。底层API也可以使用。 参考内部wiki记录。

function(op形式):   保留了fluid下面的api，一般用于底层API。
 - Accuracy:  只支持当前batch的统计，需用户做batch累积计算。底层API可以使用。
 - auc:           可计算batch和累积batch的auc。 PaddleRec中使用。

File: python/paddle/metric/metrics.py (hunk 17 above)
Comment (qingqing01) [reply python/paddle/metric/metrics.py]: @jzhang533  Precision和Recall的功能会上讨论后续的版本还需增强，见上面回复，本次先以统一为主。所以后续可以增强下。

File: python/paddle/metric/__init__.py (hunk 15 above)
Comment (jzhang533) [reply python/paddle/metric/__init__.py:24]: 计划废弃的API都添加一下`deprecated` decorator吧。

File: python/paddle/metric/metrics.py (hunk 17 above)
Comment (jzhang533) [reply python/paddle/metric/metrics.py]: 我看model.prepare的metrics是可以接受list的，这里改一下是否就可以？（先不考虑运行效率问题）

Comment (qingqing01) [reply python/paddle/metric/metrics.py]: @jzhang533  Done. Thanks!

**Review by XiaoguangHu01 — DISMISSED** [review]
LTGM

**Review by jzhang533 — DISMISSED** [review]
lgtm
will have follow up pr to deprecate fluid apis.

**Review by raindrops2sea — DISMISSED** [review]

**Review by saxon-zh — DISMISSED** [review]
LGTM

**Comment by qingqing01** [comment]
Fix  sample code about Adam, since  https://github.com/PaddlePaddle/Paddle/pull/26288 is merged. 

@XiaoguangHu01 @raindrops2sea  @jzhang533  @saxon-zh  Please help to review again.

**Review by saxon-zh — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by jzhang533 — APPROVED** [review]
lgtm
